subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/server/ServerWithExpectations.groovy

```diff
@@ -43,7 +43,7 @@ abstract class ServerWithExpectations extends ExternalResource {
         resetExpectations()
     }
 
-    protected void onFailure(Throwable failure) {
+    protected synchronized void onFailure(Throwable failure) {
```

Comment:
There are 2 different lifecycles: the setup of the test (`given`, `when`, `then`) which happens serially, and the execution of the build, with potential failures, which can happen in parallel.